### PR TITLE
mori : WRITE-mode + Wide-EP Internode sidecar support (2P2D DP=EP=16 TP=1)

### PR DESCRIPTION
**Pull request by shikamd123**
**What type of PR is this?**

feature

**What this PR does / why we need it**:

Extends the MoRI-IO WRITE-mode + Wide-EP sidecar path from **single-pod** intra-node Wide-EP (DP=EP=8, the 1P1D case in #1463) to **multi-pod cross-pod** Wide-EP, where one DP world is split across 2+ prefill and 2+ decode pods. The target topology is DeepSeek-V3 **Wide-EP-16 = DP=16, dp_local=8** across 2 prefill + 2 decode pods (2P2D, TP=1) on AMD MI300X.

This branch is a superset of #1463 and carries that work forward unchanged, then adds the multi-pod layer on top:

1. **WRITE-mode wire shape on the prefill leg.** Populate `kv_transfer_params` with `remote_host` (decode pod IP from `POD_IP`), `remote_handshake_port`, `remote_notify_port`, `remote_dp_rank` and a stable `transfer_id` so the prefill engine has the destination it needs to RDMA-Write KV blocks to decode.
2. **Decode-side `kv_transfer_params` backfill.** vLLM's `MoRIIOConnector.request_finished` does not echo back `transfer_id`, `remote_notify_port` or `remote_handshake_port`; the sidecar fills them from its static config to prevent `KeyError` on `send_notify_block`.
3. **Concurrent prefill/decode dispatch (`--moriio-parallel-dispatch`).** `runNIXLProtocolV2WriteParallel` builds both bodies up-front and fires the two legs in parallel goroutines so decode's `update_state_after_alloc` / `send_notify_block` / block allocation overlaps prefill's forward pass.
4. **Configurable `--moriio-dp-size`.** Emit `remote_dp_size` on both legs so vLLM iterates its per-DP-rank handshake loop over the configured DP world size instead of a hardcoded 1.
5. **Stable per-request DP-rank pinning.** Route both legs of a disagg pair to rank H = `blake2s(uuid) % dpSize` via the `X-Data-Parallel-Rank` header + `remote_dp_rank` / `remote_dp_rank_override` fields, neutralising the `SO_REUSEPORT` shuffle that vLLM's `--api-server-count > 1` induces.

New in this PR — **multi-pod Wide-EP fan-out** (the 2P2D layer):

6. **`--moriio-remote-hosts` / `--moriio-decode-hosts` opposite-host-list fan-out.** A single DP world now spans multiple pods, so a single `remote_host` is no longer enough. The sidecar emits an ordered `remote_hosts` array on each leg, set to the **opposite** side's pod IPs:
   - **Decode leg** gets `remote_hosts = <prefill pod IPs>` (`--moriio-remote-hosts`) so the decode engine fans its per-DP-rank handshake out across the prefill pods.
   - **Prefill leg** gets `remote_hosts = <decode pod IPs>` (`--moriio-decode-hosts`) so prefill RDMA-writes KV to the correct decode pod. Omitting it collapses the prefill leg to a single decode host and every cross-pod write hangs.
7. **`--moriio-dp-size-local` for pod-index mapping.** Emitted as `remote_dp_size_local` on both legs. vLLM maps `pod_idx = global_dp_rank / dp_size_local`, so `remote_hosts[i]` services DP ranks `[i*dp_size_local, (i+1)*dp_size_local)`. For Wide-EP-16 this is `dp_size=16`, `dp_size_local=8`, 2 hosts per side.
8. **Fail-fast `Complete()` validation (`validateWideEPHosts`).** When either host list has >1 entry, require `dp_size_local > 0`, `dp_size % dp_size_local == 0`, and `dp_size / dp_size_local == len(hosts)` — otherwise a misconfigured fan-out surfaces only as a cross-pod handshake that hangs until the deferred timeout trips. Lists of 0 or 1 host are the single-pod case and pass unchecked.

**All additions are opt-in behind new `--moriio-*` flags.** With the host-list flags unset, the wire shape is bit-identical to the 1P1D DP=EP=8 path (no `remote_hosts` / `remote_dp_size_local` keys -> vLLM's single-host branch), which is in turn bit-identical to the legacy NIXLv2 path when no `--moriio-*` flags are set. Non-MoRI-IO users (e.g. NVIDIA) and other connectors (`sglang`, `shared_storage`, `epd`) are entirely untouched.

### Files changed (10 total: 8 modified + 2 new)

| File | Δ | Purpose |
| --- | --- | --- |
| go.mod | +1 | Add golang.org/x/crypto (for blake2s) |
| go.sum | +2 | Hash entries for the new dep |
| pkg/sidecar/proxy/connector_nixlv2.go | 338 | WRITE-mode wire shape, kv_transfer_params backfill, parallel dispatch, DP-rank pinning, multi-pod `remote_hosts` / `remote_dp_size_local` fan-out |
| pkg/sidecar/proxy/connector_nixlv2_test.go | 354 | 1P1D + 2P2D dispatch/fan-out tests |
| pkg/sidecar/proxy/dp_rank.go (new) | 43 | `pickDPRank` blake2s helper |
| pkg/sidecar/proxy/dp_rank_test.go (new) | 146 | Determinism + distribution tests |
| pkg/sidecar/proxy/options.go | 113 | `--moriio-*` flag defs + `validateWideEPHosts` |
| pkg/sidecar/proxy/options_test.go (new) | 170 | `validateWideEPHosts` branch coverage + `Complete()` wiring |
| pkg/sidecar/proxy/proxy.go | 74 | Config `MoRIIO*` fields (incl. `MoRIIORemoteHosts` / `MoRIIODPSizeLocal` / `MoRIIODecodeHosts`), `X-Data-Parallel-Rank` header, kv_transfer_params field constants |
| test/sidecar/mock/chat_completions_handler.go | 63 | MoRIIOWriteMode test gate + records request headers to assert DP-rank pins |

(`git diff --stat` reports 1277 insertions, 27 deletions across the 10 files.)

### New CLI flags (all default to OFF / pre-existing behaviour)

```
# carried from the 1P1D base (#1463)
--moriio-write-mode              --moriio-parallel-dispatch
--moriio-dp-size                 --moriio-tp-size
--moriio-local-pod-ip            --moriio-decode-handshake-port
--moriio-decode-notify-port      --moriio-decode-dp-rank
--moriio-prefill-handshake-port  --moriio-prefill-notify-port
--moriio-prefill-dp-rank

# new in this PR (multi-pod Wide-EP fan-out)
--moriio-remote-hosts            # ordered prefill pod IPs  -> decode leg remote_hosts
--moriio-decode-hosts            # ordered decode pod IPs   -> prefill leg remote_hosts
--moriio-dp-size-local           # per-pod DP size; pod_idx = dp_rank / dp_size_local
```

### Companion PRs

- **vLLM-side**: the `MoRIIOConnector` consumes `remote_hosts` / `remote_dp_size_local` from `kv_transfer_params` and resolves `pod_idx = global_dp_rank // dp_size_local`, `host = remote_hosts[pod_idx]` per handshake. When those keys are absent (the 1P1D path) it falls through to the single-host branch unchanged. The `remote_dp_rank_override` sentinel still pins the rank verbatim to avoid Go `blake2s-256` vs Python `blake2s-8` hash divergence.
- **Deployment guide (`llm-d/llm-d`)**: the `guides/pd-disaggregation/.../moriio-2p2d` kustomize overlay wires these flags onto the decode sidecar (`--moriio-dp-size=16 --moriio-dp-size-local=8 --moriio-remote-hosts=<prefill master,child> --moriio-decode-hosts=<decode master,child>`) for the 2P2D DSV3 deployment.

### Test coverage

```
go test ./pkg/sidecar/proxy/...
```

- **Integration (proxy, Ginkgo)** — `connector_nixlv2_test.go`:
  - 1P1D DP=8 concurrent dispatch: both legs pinned to one DP rank (header + `remote_dp_rank`), decode flips `do_remote_prefill=true`, `remote_dp_size` emitted, shared `transfer_id`, single-pod omits `remote_hosts`.
  - 2P2D DP=EP=16 multi-pod: opposite-host-list fan-out invariant (prefill `remote_hosts` = decode IPs, decode `remote_hosts` = prefill IPs), `remote_dp_size_local` / `remote_dp_size`.
  - 1P1D DP=8 serial path: prefill header pin + decode kv backfill.
  - Backward-compat lock: flags-off keeps the legacy NIXLv2 wire shape and emits no `X-Data-Parallel-Rank` header.
- **Unit (options)** — `options_test.go`: `validateWideEPHosts` branch coverage and `Complete()` wiring for both host lists; WRITE-mode pod-IP and parallel-dispatch preconditions.
- `dp_rank_test.go`: `pickDPRank` determinism, distribution, and edge cases.
- `make test-unit-sidecar` runs the same suite under the repo's builder container.

### End-to-end production validation

| Cluster | 4x MI300X-8 nodes (2 prefill + 2 decode) |
| --- | --- |
| Model | DeepSeek-V3 |
| Topology | 2P2D multi-node Wide-EP, EP=16 / TP=1 / DP=16 (dp_local=8) |
| Sidecar image | built from this branch |
| vLLM image | built with the companion vLLM PR's commits applied |

**Which issue(s) this PR fixes**:

This PR is part of (but does not by itself fully close) two umbrella issues in `llm-d/llm-d`; both also depend on the companion vLLM PR and the `llm-d/llm-d` guide overlay.

- Part of **[P/D][AMD]: Support AMD MoRI-IO llm-d#1469** — delivers the disaggregation-sidecar half across both single-pod and multi-pod Wide-EP.
- Part of **[WideEP]: AMD MI355X MoRI EP Guide for Pollara/Thor-2 NIC llm-d#1470** — adds the router-side enabler for genuine multi-node Wide-EP (DP=EP=16 across pods): the `remote_hosts` / `remote_dp_size_local` fan-out plus stable per-request DP-rank pinning.

**Release note**:

```
Add multi-pod Wide-EP fan-out to the disaggregation sidecar, extending
MoRI-IO WRITE-mode Wide-EP from single-pod (DP=EP=8) to cross-pod
deployments (e.g. 2P2D DP=EP=16, dp_local=8).  New opt-in CLI flags:
--moriio-remote-hosts, --moriio-decode-hosts, --moriio-dp-size-local
(emitting kv_transfer_params remote_hosts / remote_dp_size_local).
Default behaviour is unchanged when no --moriio-* flags are set.
```

Signed-off-by: Shiksha Patel <shiksha.patel@amd.com>

### CODE REVIEW

**Comment by shikamd123** [comment]
@robertgshaw2-redhat  could you please take a look at the PR and suggest obvious breaks you come across? I will make update after the feedback and actively testing the wide-ep internode deployment. cc @ravi.gupta@amd.com

**Comment by shikamd123** [comment]
@robertgshaw2-redhat  could you please take an initial look and provide the feedback on this llmd PR? I will update the PR after your  feedback. cc  ravi.gupta@amd.com @Vincent.Cave@amd.com

**Comment by vMaroon** [comment]
can u rebase and sign commits?

File: pkg/sidecar/proxy/connector_nixlv2.go [hunk 1]
```diff
@@ -285,3 +351,259 @@ func (s *Server) handleNIXLV2(w http.ResponseWriter, r *http.Request, prefillPod
...
+	parentCtx := r.Context()
+	requestStartedAt := time.Now()
+
+	// Snapshot client fields before mutating completionRequest into the prefill
+	// body; they are restored when building the decode body.
+	streamValue, streamOk := completionRequest[requestFieldStream]
+	streamOptionsValue, streamOptionsOk := completionRequest[requestFieldStreamOptions]
+	maxTokensValue, maxTokensOk := completionRequest[requestFieldMaxTokens]
+	maxCompletionTokensValue, maxCompletionTokensOk := completionRequest[requestFieldMaxCompletionTokens]
+
```
[the remark this answers is not in the document]
Comment (shikamd123) [reply pkg/sidecar/proxy/connector_nixlv2.go:459]: adressed

File: pkg/sidecar/proxy/connector_nixlv2.go [hunk 2]
```diff
@@ -285,3 +351,259 @@ func (s *Server) handleNIXLV2(w http.ResponseWriter, r *http.Request, prefillPod
...
+		pkv["remote_hosts"] = hosts
+		if s.config.MoRIIODPSizeLocal > 0 {
+			pkv["remote_dp_size_local"] = s.config.MoRIIODPSizeLocal
+		}
+	}
+	completionRequest[requestFieldStream] = false
+	delete(completionRequest, requestFieldStreamOptions)
+	completionRequest[requestFieldMaxTokens] = 1
+	completionRequest[requestFieldMaxCompletionTokens] = 1
+
```
[the remark this answers is not in the document]
Comment (shikamd123) [reply pkg/sidecar/proxy/connector_nixlv2.go:500]: addressed

File: pkg/sidecar/proxy/connector_nixlv2.go [hunk 3]
```diff
@@ -285,3 +351,259 @@ func (s *Server) handleNIXLV2(w http.ResponseWriter, r *http.Request, prefillPod
...
+		completionRequest[requestFieldStreamOptions] = streamOptionsValue
+	}
+	delete(completionRequest, requestFieldMaxTokens)
+	if maxTokensOk {
+		completionRequest[requestFieldMaxTokens] = maxTokensValue
+	}
+	delete(completionRequest, requestFieldMaxCompletionTokens)
+	if maxCompletionTokensOk {
+		completionRequest[requestFieldMaxCompletionTokens] = maxCompletionTokensValue
+	}
```
[the remark this answers is not in the document]
Comment (shikamd123) [reply pkg/sidecar/proxy/connector_nixlv2.go:525]: addressed

File: pkg/sidecar/proxy/connector_nixlv2.go [hunk 4]
```diff
@@ -93,13 +114,31 @@ func (s *Server) handleNIXLV2(w http.ResponseWriter, r *http.Request, prefillPod
...
+	} else {
+		completionRequest[requestFieldKVTransferParams] = map[string]any{
+			requestFieldDoRemoteDecode:  true,
+			requestFieldDoRemotePrefill: false,
+			requestFieldRemoteEngineID:  nil,
+			requestFieldRemoteBlockIDs:  nil,
+			requestFieldRemoteHost:      nil,
+			requestFieldRemotePort:      nil,
+		}
 	}
```
[the remark this answers is not in the document]
Comment (shikamd123) [reply pkg/sidecar/proxy/connector_nixlv2.go:169]: addressed

File: pkg/sidecar/proxy/connector_nixlv2.go [hunk 5]
```diff
@@ -226,7 +273,26 @@ func (s *Server) handleNIXLV2(w http.ResponseWriter, r *http.Request, prefillPod
...
+			}
+			if _, present := dKVParams[requestFieldRemoteDPRank]; !present {
+				dKVParams[requestFieldRemoteDPRank] = pickDPRank(uuidStr, s.config.MoRIIODPSize)
+				dKVParams[requestFieldRemoteDPRankOverride] = true
+			}
+			if _, present := dKVParams[requestFieldRemoteHandshakePort]; !present {
+				dKVParams[requestFieldRemoteHandshakePort] = s.config.MoRIIODecodeHandshakePort
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (shikamd123) [reply pkg/sidecar/proxy/connector_nixlv2.go:377]: addressed

File: pkg/sidecar/proxy/connector_nixlv2.go [hunk 6]
```diff
@@ -285,3 +351,259 @@ func (s *Server) handleNIXLV2(w http.ResponseWriter, r *http.Request, prefillPod
...
+		attribute.String("llm_d.pd_proxy.connector", "nixlv2"),
+		attribute.Bool("llm_d.pd_proxy.parallel_dispatch", true),
+	)
+	preq := r.Clone(pCtx)
+	preq.Header.Set(requestHeaderRequestID, uuidStr)
+	if s.config.MoRIIODPSize > 1 {
+		preq.Header.Set(requestHeaderDataParallelRank, strconv.Itoa(dpRank))
+	}
+	preq.Body = io.NopCloser(strings.NewReader(string(pbody)))
+	preq.ContentLength = int64(len(pbody))
```
[the remark this answers is not in the document]
Comment (shikamd123) [reply pkg/sidecar/proxy/connector_nixlv2.go]: addressed

File: pkg/sidecar/proxy/connector_nixlv2.go [hunk 7]
```diff
@@ -285,3 +351,259 @@ func (s *Server) handleNIXLV2(w http.ResponseWriter, r *http.Request, prefillPod
...
+		attribute.String("llm_d.pd_proxy.connector", "nixlv2"),
+		attribute.Bool("llm_d.pd_proxy.parallel_dispatch", true),
+	)
+	dreq := r.Clone(dCtx)
+	dreq.Header.Set(requestHeaderRequestID, uuidStr)
+	if s.config.MoRIIODPSize > 1 {
+		dreq.Header.Set(requestHeaderDataParallelRank, strconv.Itoa(dpRank))
+	}
+	dreq.Body = io.NopCloser(strings.NewReader(string(dbody)))
+	dreq.ContentLength = int64(len(dbody))
```
[the remark this answers is not in the document]
Comment (shikamd123) [reply pkg/sidecar/proxy/connector_nixlv2.go]: addressed

File: test/sidecar/mock/chat_completions_handler.go [hunk 8]
```diff
@@ -125,10 +143,37 @@ func (cc *ChatCompletionHandler) ServeHTTP(w http.ResponseWriter, r *http.Reques
...
+				if _, ok := kvTransferParamsMap["remote_notify_port"]; !ok {
+					w.WriteHeader(http.StatusBadRequest)
+					w.Write([]byte("expected remote_notify_port:<int> in WRITE mode")) //nolint:all
+					return
+				}
+				if _, ok := kvTransferParamsMap["transfer_id"]; !ok {
+					w.WriteHeader(http.StatusBadRequest)
+					w.Write([]byte("expected transfer_id:<uuid> in WRITE mode")) //nolint:all
+					return
+				}
```
[the remark this answers is not in the document]
Comment (shikamd123) [reply test/sidecar/mock/chat_completions_handler.go]: addressed

**Review by vcave — APPROVED** [review]
Dormant to enable more work on dependent PRs. Thanks Shiksha!

File: pkg/sidecar/proxy/dp_rank.go [hunk 9]
```diff
@@ -0,0 +1,43 @@
+/*
+Copyright 2025 The llm-d Authors.
```
Comment (vMaroon) [inline pkg/sidecar/proxy/dp_rank.go]: new files should have 2026 as year.

Comment (shikamd123) [reply pkg/sidecar/proxy/dp_rank.go]: My bad, let me update it.

File: pkg/sidecar/proxy/options.go [hunk 10]
```diff
@@ -329,6 +396,75 @@ func (opts *Options) Complete() error {
 		return fmt.Errorf("failed to parse target URL: %w", err)
 	}
 
+	// MoRI-IO feature gate: when MoRIIOFeatureEnabled is false, reject any
+	// attempt to use MoRI-IO flags. This keeps the feature dormant until full
+	// CI and production validation is complete.
+	if !MoRIIOFeatureEnabled && opts.MoRIIOWriteMode {
```
Comment (vMaroon) [inline pkg/sidecar/proxy/options.go]: The gate only rejects --moriio-write-mode. --moriio-dp-size > 1 passes validation and sets x-data-parallel-rank on both legs in plain READ-mode NIXLv2 (connector_nixlv2.go:104-106, 309-313) — a live routing change while the feature is dormant. Suggest the gate reject any --moriio-* flag at a non-default value, which also makes the README's "flags will error if used" accurate.

File: pkg/sidecar/proxy/connector_nixlv2.go [hunk 11]
```diff
@@ -106,13 +126,51 @@ func (s *Server) handleNIXLV2(w http.ResponseWriter, r *http.Request, prefillPod
...
 	}
 
-	completionRequest[requestFieldKVTransferParams] = map[string]any{
-		requestFieldDoRemoteDecode:  true,
-		requestFieldDoRemotePrefill: false,
-		requestFieldRemoteEngineID:  nil,
-		requestFieldRemoteBlockIDs:  nil,
-		requestFieldRemoteHost:      nil,
-		requestFieldRemotePort:      nil,
+	// Snapshot the original request map before prefill mutations so the
```
Comment (vMaroon) [inline pkg/sidecar/proxy/connector_nixlv2.go]: Dead code on the hot path: this clone runs per request and is discarded; the fallback path the comment references doesn't exist on main. Same for the original capture at L59 (_ = original in the parallel func) and transferID at L75 outside WRITE mode. Please drop the clone and move the transferID construction behind the write-mode check.

File: pkg/sidecar/proxy/options.go (hunk 10 above)
Comment (shikamd123) [reply pkg/sidecar/proxy/options.go]: agree

File: pkg/sidecar/proxy/connector_nixlv2.go (hunk 11 above)
Comment (shikamd123) [reply pkg/sidecar/proxy/connector_nixlv2.go]: ok

**Comment by shikamd123** [comment]
@vMaroon  as noted earlier , The coverage drop is expected **due to dormant feature code**
The code will be tested once the feature is enabled soon.

**Review by vMaroon — APPROVED** [review]
Thanks @shikamd123!

To set context for other reviewers, we're expecting quick iterations to land the functional AMD guide for this llm-d release this week. Taking this "configuration gated" work in to make sure it's included.
